src/engine/bpmcontrol.cpp

```diff
@@ -850,7 +850,7 @@ void BpmControl::collectFeatures(GroupFeatureState* pGroupFeatures) const {
                        &dThisBeatLength, &dThisBeatFraction)) {
         pGroupFeatures->has_beat_length_sec = true;
         // Note: dThisBeatLength is fractional frames count * 2 (stereo samples)  
-        pGroupFeatures->beat_length_sec = dThisBeatLength / m_pTrack->getSampleRate() / 2 * calcRateRatio();
+        pGroupFeatures->beat_length_sec = dThisBeatLength / m_pTrack->getSampleRate() / 2 / calcRateRatio();
```

Comment:
Is this division safe, i.e. is calcRateRatio() != 0? Please add a debug assertion before the division, just in case.

Reply:
it can be zero or negative

Reply:
Since #1662 was merged, the ratio ratio could be zero or negative, which is nonsensical. I don't know what should be done in that situation. I propose reverting #1662.

Reply:
Hm, I missed 29ef9d1a708a1937dfaef90d2df33b43cdb718a6 -- before that it was also possible for the rate slider to cause the deck stop or go in reverse and we had to code defensively against that. I don't think the idea behind #1662 is flawed, just the ability for it to cause the speed to go below zero. I think a `std::max(1e-6, rateRatio)` in calcRateRatio would be reasonable.

Reply:
I tried `std::max(1e-6, rateRatio)`. That prevents `visual_bpm` from going below 0, but playback and the waveforms can still stop and reverse if `rateRange` > 1. I think the best solution at this point is to revert #1662, for which I have reopened #1768. I still think it was inappropriate to rush such a drastic change into a stable branch without extensive testing. We have not had users report bugs related to it because the maximum rate range option presented in the preferences is still 90%. I had to set `rateRange` manually with the Developer Tools window to test changes on this branch. I don't dispute that there is a use case for allowing `rateRange` to go above 100%, but that would be a new feature which deserves thorough review before merging to the master branch and going through a full cycle of beta testing, not rushed into a stable branch.

Reply:
Ok, SGTM. Though it'd be good to be defensive here and avoid a division by zero and negative values. Even with #1662 reverted calcRateRatio can return `<=0` if a controller script sets `rate` out of bounds (it has `allowOutOfBounds` set).

Reply:
`rate_dir` also has no limits on what value it can be set to

Reply:
Good points. I added the `std::max`. I still think it would be good to merge #1768.